[GLib] Add WebSetting for media-session support

philn opened the pull request:
#### d9e28adf8b1a015072ab578f96961f1b9df90a8a
<pre>
[GLib] Add WebSetting for media-session support
<a href="https://bugs.webkit.org/show_bug.cgi?id=250269">https://bugs.webkit.org/show_bug.cgi?id=250269</a>

Reviewed by NOBODY (OOPS!).

The MediaSession support (which also exposes media elements over MPRIS) is disabled by default at
runtime because not ready for prime time just yet. See also
<a href="https://bugs.webkit.org/show_bug.cgi?id=247527.">https://bugs.webkit.org/show_bug.cgi?id=247527.</a>

The PlatformMediaSession code is independant from the MEDIA_SESSION support so in order to enable/disable
DBus notifications depending on the MediaSession preference, a `webcoreOnChange` callback was added
for this purpose in SettingsBase.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::mediaSessionEnabledChanged):
* Source/WebCore/page/SettingsBase.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webKitSettingsConstructed):
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_enable_media_session):
(webkit_settings_set_enable_media_session):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9e28adf8b1a015072ab578f96961f1b9df90a8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11690 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35592 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111826 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172046 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106527 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2593 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94835 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109535 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9732 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37392 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91592 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79139 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92802 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5148 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25868 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89124 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2829 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5308 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2322 "Found 1 new test failure: webgl/2.0.0/conformance2/textures/canvas_sub_rectangle/tex-3d-rgb8-rgb-unsigned_byte.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29624 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45359 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92050 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7040 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20625 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
```diff
@@ -530,6 +530,13 @@ WEBKIT_API void
+
+WEBKIT_API void
+webkit_settings_set_enable_media_session                       (WebKitSettings *settings,
+                                                                gboolean enabled);
```
Alignment is wrong here

philn replied:
Are you talking about the second argument? I used same alignment as above.

mcatanzaro replied:
Should be:

```suggestion
                                                                gboolean        enabled);
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -3964,3 +3983,46 @@ void webkit_settings_set_media_content_types_requiring_hardware_support(WebKitSe
+    priv->preferences->setMediaSessionCoordinatorEnabled(enabled);
+    priv->preferences->setMediaSessionPlaylistEnabled(enabled);
+#endif
+    g_object_notify_by_pspec(G_OBJECT(settings), sObjProperties[PROP_ENABLE_MEDIA_SESSION]);
```
This will notify twice if setting the property directly, but I notice all other settings have the same bug, so no need to change anything here: might as well be consistent with the other settings.

If we ever find a 25 hour day, we should do an audit to see where else we are missing G_PARAM_EXPLICIT_NOTIFY.

mcatanzaro reviewed (approved):
Seems good.

philn committed:
[GLib] Add WebSetting for media-session support
https://bugs.webkit.org/show_bug.cgi?id=250269

Reviewed by NOBODY (OOPS!).

The MediaSession support (which also exposes media elements over MPRIS) is disabled by default at
runtime because not ready for prime time just yet. See also
https://bugs.webkit.org/show_bug.cgi?id=247527.

The PlatformMediaSession code is independant from the MEDIA_SESSION support so in order to enable/disable
DBus notifications depending on the MediaSession preference, a `webcoreOnChange` callback was added
for this purpose in SettingsBase.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::mediaSessionEnabledChanged):
* Source/WebCore/page/SettingsBase.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webKitSettingsConstructed):
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_enable_media_session):
(webkit_settings_set_enable_media_session):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):

calvaris reviewed (commented):
LGTM

carlosgcampos commented on Source/WebCore/page/SettingsBase.cpp:
```diff
@@ -437,6 +441,19 @@ void SettingsBase::layerBasedSVGEngineEnabledChanged()
+    bool enabled = false;
+    if (m_page)
+        enabled = m_page->settings().mediaSessionEnabled();
+    auto& sessionManager = reinterpret_cast<MediaSessionManagerGLib&>(PlatformMediaSessionManager::sharedManager());
+    sessionManager.setDBusNotificationsEnabled(enabled);
```
I don't think a web setting is the right thing here, because media session manager is global. I see there's an existing setting for media session, but it doesn't seem to be used, or I don't find where it's used.

philn replied:
What this code does is enabling MPRIS when the media session setting is enabled, and disable MPRIS when the media session setting is disabled.

philn replied:
I get your point though, if we create one webview (A), disable media-session on it, then create another webview (B) then enable media-session on it, then MPRIS notifications will be created for media elements of A too... Not sure what to do about this.

mcatanzaro replied:
Good catch. We could either (a) make MediaSessionManager non-global, or (b) just don't add the setting?

We probably don't really _need_ the setting, do we? We've agreed it's not fully baked yet and shouldn't be exposed to users yet. Developers can enable the feature manually if desired?

philn replied:
https://github.com/WebKit/WebKit/pull/8515

sonnyp replied:
I'm a bit late here I guess but the setting would have been helpful for Tangram since tabs don't share `WebKitSettings`, a user could enable it for a specific tab.

philn replied:
We could have a websetting for media-session, but for MPRIS it would not be so easy due to the current singleton approach, I'm afraid.
